// nestjs-inertia/dist/test/index.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const testing_1 = require("@nestjs/testing");
const request = require("supertest");
const test_module_1 = require("./utils/test.module");
const template_1 = require("../src/template");
describe('test', () => {
    let server;
    beforeAll(async () => {
        const module = await testing_1.Test.createTestingModule({
            imports: [test_module_1.TestModule],
        }).compile();
        const app = module.createNestApplication();
        await app.init();
        server = app.getHttpServer();
    });
    afterAll(() => {
        server.close();
    });
    describe('1+1', () => {
        it('should pass', () => {
            expect(1 + 1).toBe(2);
        });
    });
    describe('should render view template with inertia page data', () => {
        it('GET /users', () => {
            const dataPage = JSON.stringify({
                component: 'Users/AllUser',
                props: {
                    users: [{
                            id: 1,
                            name: 'John Doe'
                        }]
                },
                url: '/users',
                version: '1'
            });
            return request(server)
                .get('/users')
                .expect(200)
                .expect((0, template_1.default)(dataPage));
        });
    });
    describe('should render view template with inertia shared and page data', () => {
        it('GET /users/2', () => {
            const dataPage = JSON.stringify({
                component: 'Users/DetailUser',
                props: {
                    isLoggedIn: true,
                    id: 2,
                    name: 'John Doe'
                },
                url: '/users/2',
                version: '1'
            });
            return request(server)
                .get('/users/2')
                .expect(200)
                .expect((0, template_1.default)(dataPage));
        });
    });
    describe('should return a redirection', () => {
        it('DELETE /users/2', () => {
            return request(server)
                .delete('/users/2')
                .expect(303);
        });
    });
});
